### ISSUE TITLE
Public API to preempt tf.train.Server

### ISSUE THREAD

**Issue by superbobry** [tensorflow/tensorflow#21492]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: N/A
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: N/A
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: N/A
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: v1.9.0-0-g25c197e023 1.9.0
- **Python version**: N/A
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A
- **Exact command to reproduce**: N/A

### Describe the problem

`tf.train_and_evaluate` uses `time.sleep` to (optimistically) synchronize the startup of chief/worker nodes in the distributed mode (see [estimator/training.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/estimator/training.py#L753)). The implicit assumption in this logic is that by the time the worker is spawned, the chief has already started its `tf.train.Server`, i.e. the scheduler should be aware of the assumption and should schedule and initialize the chief first. This might not be easily achievable on general purpose systems like YARN.

One possible solution to this is to synchronize the chief/worker tasks on a barrier, and then preemptively start the server right after the barrier, but prior to calling `train_and_evaluate`. This does not eliminate the race condition entirely but makes it much less likely in practice. The only problem here is that `train_and_evaluate` does not provide a documented way to account for preempted servers. The undocumented way is:

https://github.com/tensorflow/tensorflow/blob/2b4fd1c2b7a37367e61bbae3d27d194a894cb7bb/tensorflow/python/estimator/training.py#L747-L748

I was wondering if it would be possible to make this part of the `train_and_evaluate` contract public or, alternatively, address the issue in another way?

**Comment by karmel** [comment]
@xiejw @ispirmustafa -- what is the recommended way to handle pre-empted Servers for systems like YARN?

**Comment by superbobry** [comment]
@xiejw thank you for the clarification, and apologies for poor wording on my side. The code fragment I've linked to gives the "delay starting" rationale in the comment. 

> All workers (including chief) will try to find each other first 
Would it be right to say that in order for a task to start, it should successfully connect to all other tasks; and that if the connection is unsuccessful, the task will retry indefinitely/"long enough"? I would very much appreciate it if you could reference relevant code in the reply.

**Comment by xiejw** [comment]
Assume device_filter is not set. The precise sequence can be described as follows:

1. start_std_server. This opens the port and listens the grpc in background thread. All workers must do this first. Otherwise it might delay other workers to train the model.

2. (optional) time.sleep for delay starting. Note that at this time the grpc port is open already.

3. Estimator.train. This creates the graph and launches a new tf.Session to train the model. The session will try to talk to each other device (in this case other workers) and then run the TF graph. This is the part it will try to block until all devices are found. The code is [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/distributed_runtime/master.cc#L215).

Say there is a worker in the time.sleep for 100000 secs (basically forever). As the port has been opened in step 1, so other workers can start train (step 3) without waiting. 

With device filter, only step 3 is slightly different where each worker only tries to find the devices in the list rather than all devices.

**Comment by superbobry** [comment]
This is indeed a non-issue, thanks again @xiejw!

**Comment by superbobry** [comment]
@xiejw I thought about this more over a weekend and came up with another use-case for preemption: port reservation. 

YARN does not manage the network on the containers, so in order to run TF on YARN one has to manually reserve a port on each of the allocated containers, communicate it to all other tasks to assemble a `ClusterSpec` and only then start the server. IIUC with the current API there is no race-free way of doing this is in TF. Could you advise?

**Comment by xiejw** [comment]
@karmel do we have any expert for YARN? I do not understand that environment, so it is very hard for me to give best suggestions here. 

@superbobry As I mentioned above, I do not understand YARN. So, my suggestions could be wrong or sub-optimal. The minimal knowledge each training job should know is 

1. The port itself uses
2. The ports used by PS

The ports of other workers are optional. You still need to fill the TF_CONFIG, but the values of other workers do not matter. 

If the conditions above are met in YARN, then you can construct a device_filter (see [example code](https://github.com/tensorflow/tensorflow/blob/r1.10/tensorflow/python/estimator/run_config.py#L518)). The session will try to find the device in that list only even other workers are online (or even the other worker's addresses are fake).

**Comment by superbobry** [comment]
@xiejw the problem with YARN is that it does not manage the network, i.e. there is no way to reserve a port prior to submitting an application. One possible solution to this is to

1. Spawn a Python process in each YARN container.
2. Bind to port 0.
3. Communicate the port assigned by the OS to all other containers and aggregate a cluster spec
4. Export the spec in `TF_CONFIG` and start training.

which roughly translates to

```python
with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
    sock.bind(("", 0))
    ipaddr, port = sock.getsockname()
    cluster_spec = broadcast_addr_and_aggregate_spec(f"{ipaddr}:{port}")

# export TF_CONFIG
tf.estimator.train_and_evaluate(...)
```

Note that this implementation has a race condition between closing `sock` and binding a `tf.train.Server` to the reserved port. This means that a task from another distributed TF training running on the same YARN node could hijack the port, and effectively talk to a completely unrelated task. The window of opportunity for the race condition to occur is undefined and depends on the implementation of `train_and_evaluate`. 

I see multiple potential ways of getting rid of the race condition by slightly modifying the existing Python/C++ API of `tf.train.Server`:

* Add an API to spawn a server using an existing FD. The gRPC backend seems to already support this (see grpc/grpc#6610). 
* Allow `SO_REUSEPORT` when creating a server. Currently the option is [explicitly disabled](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc#L191).

**Comment by superbobry** [comment]
@mrry could you comment on the flakiness caused by SO_REUSEPORT?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @karmel: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by karmel** [comment]
@jhseu -- I hear you might be able to advise on YARN?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @karmel: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by jhseu** [comment]
I don't have much personal experience with running on YARN, but the LinkedIn folks do:
https://github.com/linkedin/TonY
https://engineering.linkedin.com/blog/2018/09/open-sourcing-tony--native-support-of-tensorflow-on-hadoop

**Comment by tensorflowbutler** [comment]
Nagging Assignee @karmel: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @karmel: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @karmel: It has been 47 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by karmel** [comment]
[DistributionStrategies](https://github.com/tensorflow/community/blob/master/rfcs/20181016-replicator.md) is the new API for handling distribution and synchronization, and that API will be preferred in the coming months to the tf.train.Server API. That said, we have now started a Special Interest Group (SIG) for networking issues like these, so while I am closing this issue as it pertains to tf.train.Server, I would encourage you to join and participate in the [Networking SIG](https://groups.google.com/a/tensorflow.org/forum/#!forum/networking) to ensure these questions are addressed for YARN and other systems.
